src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Hover/DefaultRazorHoverInfoService.cs

```diff
@@ -0,0 +1,167 @@
+                containingTagNameToken.Span.IntersectsWith(location.AbsoluteIndex))
+            {
+                // Hovering over HTML tag name
+                var stringifiedAttributes = _tagHelperFactsService.StringifyAttributes(attributes);
+                var binding = _tagHelperFactsService.GetTagHelperBinding(
+                    tagHelperDocumentContext,
+                    containingTagNameToken.Content,
+                    stringifiedAttributes,
+                    parentTag: null,
+                    parentIsTagHelper: false);
+
+                if (binding is null)
+                {
+                    // No matching tagHelpers, it's just HTML
+                    return null;
+                }
+                else
+                {
+                    Debug.Assert(binding.Descriptors.Count() > 0);
+
+                    var result = ElementInfoToHover(binding.Descriptors, position);
+                    return result;
+                }
+            }
+
+            if (_htmlFactsService.TryGetAttributeInfo(parent, out containingTagNameToken, out var selectedAttributeName, out attributes) &&
+                attributes.Span.IntersectsWith(location.AbsoluteIndex))
+            {
+                // Hovering over HTML attribute name
+                var stringifiedAttributes = _tagHelperFactsService.StringifyAttributes(attributes);
+
+                var binding = _tagHelperFactsService.GetTagHelperBinding(
+                    tagHelperDocumentContext,
+                    containingTagNameToken.Content,
+                    stringifiedAttributes,
+                    parentTag: null,
+                    parentIsTagHelper: false);
+
+                var tagHelperAttributes = _tagHelperFactsService.GetBoundTagHelperAttributes(tagHelperDocumentContext, selectedAttributeName, binding);
+                var attributeHoverModel = AttributeInfoToHover(tagHelperAttributes, position);
+
+                return attributeHoverModel;
+            }
+
+            return null;
+        }
+
+        private HoverModel AttributeInfoToHover(IEnumerable<BoundAttributeDescriptor> descriptors, Position position)
+        {
+            var descriptionInfos = descriptors.Select(d => new TagHelperAttributeDescriptionInfo(d.DisplayName, d.GetPropertyName(), d.TypeName, d.Documentation))
+                .ToList()
+                .AsReadOnly();
+            var attrDescriptionInfo = new AttributeDescriptionInfo(descriptionInfos);
+
+            if(!_tagHelperDescriptionFactory.TryCreateDescription(attrDescriptionInfo, out var markdown))
+            {
+                return null;
+            }
+
+            var hover = new HoverModel {
+                Contents = new MarkedStringsOrMarkupContent(new MarkupContent
+                {
+                    Kind = MarkupKind.Markdown,
+                    Value = markdown
+                }),
+                Range = new RangeModel(start: position, end: position)
```

Comment:
Trying to understand how hover works. What is this Range actually for? Does it highlight that range when you hover?

Reply:
A good question to ask, I had just kind of preceded blindly here: https://github.com/OmniSharp/csharp-language-server-protocol/blob/master/src/Protocol/Models/Hover.cs#L18-L19. You basically got it right on the first try. If we think that highlighting just the one character is a bad idea we could just not supply a Range at all.

Reply:
It should be pretty straightforward to obtain the range of the attribute name. If not, I'm okay with not supplying a range. Also, I think what you have here will probably not highlight anything because the range is zero characters.